README.md

```diff
@@ -96,6 +96,53 @@ status:
 ...
 ```
 
+## End to end Testing
+
+There is two type of e2e tests launched on CI.
+
+The first one would just apply the yaml files making sure they don't have any
+syntax issues. Pretty simple one, it just basically check the syntax.
+
+The second one would do some proper functional testing, making sure the task
+actually **ran** properly.
+
+The way the functional tests works is that if you have a directory called
+`tests/` inside the task, it would start creating a random `Namespace`, apply
+the task and then every yaml files that you have in that `tests/` directory.
+
+Usually in these other yaml files you would have a yaml file for the
+test resources (`PipelineResource`) and a yaml files for to run the task
+(`TaskRun`).
+
+Sometime you may need to be able to launch some scripts before applying the
+tested task or the other yaml files. Some may pre-setup something on the
+Namesapce or have to do something externally or sometime you may even want to do
+some manipulation of the main `Task`.
+
+For example on the *image builders* tasks like `kaniko` or `jib` we want to
+upload the tasks to a registry to make sure it is actually built properly. To do
+so we [manipulate](kaniko/tests/apply-task-hook.sh) with a python script the
+`Task` (something we don't want for everyone but only for the tests) to add a
+registry as a `Sidecar` and make sure that the `TaskRun` set the parameters to
+upload there. Simple and straightforward no need to upload to an external image
+registry provider having to setup the tokens and deals with the side effects...
+
+There is two different scripts that are checked if present in the `scripts`,
+those scripts actually sourced via the `source` bash script, so you can output
+some environment variables to it that would be applied :
+
+1. **apply-task-hook.sh**: Script to run before applying the task
+2. **apply-taskrun-hook.sh**: Script to run before applying the taskruns or other yaml files.
```

Comment:
A few questions:

1. Can we get `post-` apply actions to validate outcomes?
2. What environment are these tests running in? What tools will we have available (`kubectl`, `docker`, etc...)?
3. Are we restricted to use `bash` vs something like `go`?

For reference, these are the current integration tests we have for buildpacks: 
https://github.com/buildpack/tekton-integration

We'd be happy to move the tests to the catalog if possible.

Reply:
@chmouel Thanks for the reply.

1. Sounds good. We can certainly add it when we try in incorporate our tests into this repo.
2. Thanks! Maybe this should be documented since it's fairly critical to what you can do in your hooks.
3. This works for us and based on (2) the current test-runner image would allow for that.

Reply:
great, thank you! I have documented it,